cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/StructuredRecordToCubeFactTransform.java

```diff
@@ -0,0 +1,388 @@
+import co.cask.cdap.templates.etl.api.StageConfigurer;
+import co.cask.cdap.templates.etl.api.Transform;
+import co.cask.cdap.templates.etl.api.TransformContext;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.gson.Gson;
+
+import java.nio.ByteBuffer;
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.Collection;
+import javax.annotation.Nullable;
+
+/**
+ * Transforms a {@link StructuredRecord} into a {@link CubeFact} object that can be written to a
+ * {@link co.cask.cdap.api.dataset.lib.cube.Cube} dataset.
+ * <p/>
+ * This transformation requires a configuration for mapping {@link StructuredRecord} into a {@link CubeFact} that is
+ * provided in JSON form in {@link #MAPPING_CONFIG_PROPERTY} property. Example of the configuration:
+<pre>
+  {
+    timestamp: {
+      sourceField: "timeField",
+      sourceFieldFormat: "HH:mm:ss"
+    }
+
+    tags: [
+      {
+        name: "tag1",
+        sourceField: "field1"
+      },
+      {
+        name: "tag2",
+        value: "staticValue"
+      }
+    ],
+
+    measurements: [
+      {
+        name: "metric1",
+        type: "COUNTER",
+        sourceField: "field7"
+      },
+      {
+        name: "metric2",
+        type: "GAUGE",
+        sourceField: "field7"
+      },
+      {
+        name: "metric3",
+        type: "COUNTER",
+        value: "1"
+      },
+    ],
+  }
+</pre>
+ *
+ * In general, the value for a timestamp, tag or measurement fields can be retrieved from {@link StructuredRecord}
+ * field using 'srcField' property or set to a fixed value using 'value' property. If both are specified, then the one
+ * in 'value' is used as the default when 'srcField' is not present in {@link StructuredRecord}.
+ *
+ * <h3>Special cases</h3>
+ * <p/>
+ * Timestamp is either retrieved from the record field or is set to a current ts (at processing) is used.
+ * If record field is used, either dateFormat is used to parse the value, or it is assumed as epoch in ms.
+ * To use current ts, configure it the following way:
+ *
+<pre>
+  {
+    timestamp: {
+      value: "now"
+    },
+    ...
+  }
+ </pre>
+ *
+ * <p/>
+ * Measurement type (specified in 'type' property) can be one of {@link MeasureType} values.
+ *
+ */
+public class StructuredRecordToCubeFactTransform extends Transform<StructuredRecord, CubeFact> {
+  public static final String MAPPING_CONFIG_PROPERTY = "mapping.config";
+
+  private CubeFactBuilder factBuilder;
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(StructuredRecordToCubeFactTransform.class.getSimpleName());
+    configurer.setDescription("Transforms a StructuredRecord to a CubeFace object used to write to a Cube.");
+    configurer.addProperty(new Property(MAPPING_CONFIG_PROPERTY, "the StructuredRecord to CubeFact mapping configuration.", true));
```

Comment:
this will need more description, or at least a reference to where somebody can find out more.